### PR TITLE
Avoid out-of-memory errors when uploading a recording

### PR DESCRIPTION
**Pull request by mishamosher**
Currently, the recording server (and the backend server) loads the entirety of the recording file into RAM, which is far from optimal.

This PR changes such behavior and uses a streaming POST request instead.

# For reviewers:

I've noted that `newFile($fileName, $content)` ([source code](https://github.com/nextcloud/server/blob/215aef3cbdc1963be1bb6bca5218ee0a4b7f1665/lib/public/Files/Folder.php#L113)) accepts `string|resource|null`, and `resource` is used in this case.

In my tests it performs a *copy* (and not a simple file rename), which temporarily wastes storage space in my current implementation.

I'm unsure how to force it to perform a file rename instead, or even if the current implementation (copy) is the desired solution so that compatibility with streaming-based file storage is not affected.

### CODE REVIEW

File: lib/Service/RecordingService.php [hunk 1]
```diff
@@ -156,17 +167,23 @@ public function getContentFromFileArray(array $file, Room $room, Participant $pa
 			throw new InvalidArgumentException('invalid_file');
 		}
 
-		$content = file_get_contents($file['tmp_name']);
-		unlink($file['tmp_name']);
```
Comment (danxuliu) [inline lib/Service/RecordingService.php:160]: Just to be fully sure, `$file['tmp_name']` is automatically removed by PHP and therefore there is no need to unlink it like it was done before, right?

File: lib/Service/RecordingService.php [hunk 2]
```diff
@@ -156,17 +167,23 @@ public function getContentFromFileArray(array $file, Room $room, Participant $pa
...
+		$resource = fopen($file['tmp_name'], 'r');
+		$resourceStat = fstat($resource);
```
Comment (danxuliu) [inline lib/Service/RecordingService.php:171]: `fopen` and `fstat` can return `false` in case of failure. Although it should not happen I think it would be good to check if they failed and throw an exception in that case.

File: recording/src/nextcloud/talk/recording/BackendNotifier.py [hunk 3]
```diff
@@ -59,16 +61,26 @@ def doRequest(backend, request, retries=3):
...
-
-    if config.getBackendSkipVerify(backend):
-        context = ssl.create_default_context()
-        context.check_hostname = False
-        context.verify_mode = ssl.CERT_NONE
+    backendSkipVerify = config.getBackendSkipVerify(backend)
 
     try:
-        urlopen(request, context=context)
+        if isinstance(request, UrllibRequest):
```
Comment (danxuliu) [inline recording/src/nextcloud/talk/recording/BackendNotifier.py:69]: Rather than using both `Request` from `urllib` and `requests`, would it be possible to convert the `urllib` requests to use `requests` instead? I guess the `urllib` requests were kept just because it was the existing code (and in general I would prefer that :-) ), but in this case I think a unified approach would be better. Or is there any benefit in using `Request` from `urllib` in addition to the one from `requests`?

File: recording/src/nextcloud/talk/recording/BackendNotifier.py [hunk 4]
```diff
@@ -59,16 +61,26 @@ def doRequest(backend, request, retries=3):
     :param backend: the backend to send the request to.
     :param request: the request to send.
     :param retries: the number of times to retry in case of failure.
+    :param library: the library to use for the request. Possible values are: 'urllib', 'requests'.
```
Comment (danxuliu) [inline recording/src/nextcloud/talk/recording/BackendNotifier.py:64]: `library` seems to be a leftover from an older implementation that can be removed, as it is not explicitly given and the request type is detected via `isinstance`.

File: recording/src/nextcloud/talk/recording/BackendNotifier.py [hunk 5]
```diff
@@ -59,16 +61,26 @@ def doRequest(backend, request, retries=3):
...
+
+            if backendSkipVerify:
+                context = ssl.create_default_context()
+                context.check_hostname = False
+                context.verify_mode = ssl.CERT_NONE
+
+            urlopen(request, context=context)
+        elif isinstance(request, RequestsRequest):
+            session = Session()
+            prepped = session.prepare_request(request)
```
Comment (danxuliu) [inline recording/src/nextcloud/talk/recording/BackendNotifier.py:80]: Although `prepped` is fine, I would prefer to name it `preparedRequest` :-)

**Review by danxuliu — CHANGES_REQUESTED** [review]
Thanks a lot for your contributions and sorry for the delay.

Regarding the copy instead of the rename of the uploaded file I am afraid that I do not know, although from a quick look to [the DAV code](https://github.com/nextcloud/server/blob/98ed72b3ed7e81a75d9a323c70a5e7f5af265a23/apps/dav/lib/Connector/Sabre/File.php#L138) it seems that a copy is done also in that case (except in the case of chunked files, in which the assembled file is moved, but if I recall correctly that assembled file is nevertheless copied from the different chunks). Thus I would assume that, even if it would be better to do a move rather than a copy when possible, a copy is needed in all cases (but please note that it is just a guess).

Besides the comments, could you rebase into latest master? In Nextcloud Talk we prefer rebasing the branch and force pushing it (`--force-with-lease` :-) ) rather than merging the master branch into the pull request branch.

Could you also squash the commit that updates the copyright into the commit that modified the file in first place, the commits to adjust the tests into the commit that modified the code under test, and the commit that fixes the formatting into the commits that added the wrong formatted code?

Thanks!

File: lib/Service/RecordingService.php (hunk 1 above)
Comment (mishamosher) [reply lib/Service/RecordingService.php:160]: The [official behavior](https://www.php.net/manual/en/features.file-upload.post-method.php) is:

> The file will be deleted from the temporary directory at the end of the request if it has not been moved away or renamed.

In my implementation, neither of those actions (move/rename) is performed on the original temporary file, so it is guaranteed to be removed after the PHP script ends.

An additional `unlink` could be added after [this line](https://github.com/nextcloud/spreed/pull/9253/files#diff-b8455de898ee16a5504ba2177a4674041bf91734999d5405b8ca9cba40cf7eaeR136), but I'm not sure if all the implementations of `file->newFile(...)` guarantee that the file is fully copied after the function call ends.

File: lib/Service/RecordingService.php (hunk 2 above)
Comment (mishamosher) [reply lib/Service/RecordingService.php:171]: I'm not that familiar with the totality of the spreed/Talk source code. Is there a recommended exception that should be used in those cases, or the built-in ones should suffice (UnexpectedValueException)?

File: recording/src/nextcloud/talk/recording/BackendNotifier.py (hunk 3 above)
Comment (mishamosher) [reply recording/src/nextcloud/talk/recording/BackendNotifier.py:69]: From the docs, I get the impression that `requests` is more oriented towards `functional programming` and `urllib` towards `object-oriented programming`, although `object-oriented programming` is not impossible to achieve with `requests`. Kinda like PHP 4/5 vs PHP 7/8 differences.

Lately `requests` has been moving towards an `object-oriented` approach, so this difference is slowly becoming a non-issue.

If removing `urllib` entirely is not an issue, would it be acceptable to add a separate PR for such change, and keep the current implementation for this PR?

File: lib/Service/RecordingService.php (hunk 1 above)
Comment (danxuliu) [reply lib/Service/RecordingService.php:160]: > In my implementation, neither of those actions (move/rename) is performed on the original temporary file, so it is guaranteed to be removed after the PHP script ends.

Sounds good :-)
 
> An additional `unlink` could be added after [this line](https://github.com/nextcloud/spreed/pull/9253/files#diff-b8455de898ee16a5504ba2177a4674041bf91734999d5405b8ca9cba40cf7eaeR136), but I'm not sure if all the implementations of `file->newFile(...)` guarantee that the file is fully copied after the function call ends.

I am not aware of any implementation that does an asynchronous copy. But if there were, I guess it would fail anyway due to the temporal file being deleted at the end of the request. So in practice I think that the implicit unlink is fine (and if it turns out that left over files are eventually found we can always fix it later).

File: lib/Service/RecordingService.php (hunk 2 above)
Comment (danxuliu) [reply lib/Service/RecordingService.php:171]: The rest of the method just uses generic `InvalidArgumentException` with a descriptive parameter, so that could be (ab)used again in this case :-) (even if it does not exactly match [the official description of the exception](https://www.php.net/manual/en/class.invalidargumentexception.php))

File: recording/src/nextcloud/talk/recording/BackendNotifier.py (hunk 3 above)
Comment (danxuliu) [reply recording/src/nextcloud/talk/recording/BackendNotifier.py:69]: > From the docs, I get the impression that `requests` is more oriented towards `functional programming` and `urllib` towards `object-oriented programming`, although `object-oriented programming` is not impossible to achieve with `requests`. Kinda like PHP 4/5 vs PHP 7/8 differences.
> 
> Lately `requests` has been moving towards an `object-oriented` approach, so this difference is slowly becoming a non-issue.

I am not familiar enough neither with `requests` nor `urllib`, so I can not provide any insight on that, sorry :-)

> If removing `urllib` entirely is not an issue, would it be acceptable to add a separate PR for such change, and keep the current implementation for this PR?

Yes, it would be fine. But I thought that it would just require changing `backendRequest = UrllibRequest(url, data, headers)` to `backendRequest = Request('POST', url, headers, data)` (based on the change done in `uploadRecording`) :thinking: Is something else required, or why would you like to do it in a different pull request? For the record, my idea was to first move the existing `urllib` code to `requests` and then, in a separate commit, adjust it to stream the upload.

File: lib/Service/RecordingService.php (hunk 1 above)
Comment (mishamosher) [reply lib/Service/RecordingService.php:160]: Will keep the current implementation then (:

File: lib/Service/RecordingService.php (hunk 2 above)
Comment (mishamosher) [reply lib/Service/RecordingService.php:171]: Got it. Thanks.

File: recording/src/nextcloud/talk/recording/BackendNotifier.py (hunk 3 above)
Comment (mishamosher) [reply recording/src/nextcloud/talk/recording/BackendNotifier.py:69]: After reading all of the `urllib` calls, all of them are `POST` requests. Completely replacing `urllib` by `requests` should require minimal changes as you suggest.

Asked about a separate PR because of the description, but that can be adjusted anyways.

tl;dr if completely replacing `urllib` by `requests` is desired, so be it in this very PR (:

**Comment by mishamosher** [comment]
@danxuliu done with most (if not all) of the requested changes.

File: recording/src/nextcloud/talk/recording/BackendNotifier.py (hunk 3 above)
Comment (danxuliu) [reply recording/src/nextcloud/talk/recording/BackendNotifier.py]: > After reading all of the `urllib` calls, all of them are `POST` requests. Completely replacing `urllib` by `requests` should require minimal changes as you suggest.

He, now you made me wonder about what would happen if non-POST requests are needed at some point. But... let's worry about that in the future if needed ;-)

> Asked about a separate PR because of the description, but that can be adjusted anyways.

In general I prefer separate pull requests that are rebased one on top of the other (to nickvergessen's dismay :-P ), but in this case I think it is fine to do the groundwork of moving from `urllib` to `requests` as a separate commit in the same pull request.

> tl;dr if completely replacing `urllib` by `requests` is desired, so be it in this very PR (:

Perfect, thanks :-)

**Review by danxuliu — COMMENTED** [review]
Thanks!

Would it be possible to split the first commit in one commit that changes from `urllib` to `requests`, and then in another commit to use streaming POST requests? If moving to `requests` requires using `MultipartEncoder` and, with that, it implicitly provides streaming POST requests, or if `encode_multipart_formdata` can not be used with `requests` and something complex is needed that will be just replaced by `MultipartEncoder` in the next commit then it is fine in a single commit, of course. But if possible I would prefer it to be split to have more "self-contained" commits.

Regarding the PHP changes I would prefer if the tests are kept in sync with the code under test even at the commit level, so I would squash the third commit into the second one.

All this is nitpicking (sorry!), so I can do the changes myself if you want, just let me know :-)

**Comment by mishamosher** [comment]
@danxuliu changes applied.

**Review by danxuliu — APPROVED** [review]
Tested and works :+1:

I took the liberty to adjust the second commit, as it was using the already removed `encode_multipart_formdata`. I hope that is fine :-)

Independently of that, excellent work! Thanks!

**Comment by mishamosher** [comment]
The requests library uses urllib as a dependency, so it was not really
removed.

To be honest, I'm just hoping to have the recording functionality working
out-of-the-box in my setup (:

Thanks to you for taking the time to review the PR.

**Comment by danxuliu** [comment]
> The requests library uses urllib as a dependency, so it was not really removed.

Argh, you are right... I forgot that it was a dependency. I meant removed from the imports, but it could have been added back to the imports instead of using already the `MultipartEncoder`. But I enabled auto-merge and now it is too late :facepalm: 

> To be honest, I'm just hoping to have the recording functionality working out-of-the-box in my setup (:

We are almost there :-) I hope... :-P

> Thanks to you for taking the time to review the PR.

Thanks to you for applying the changes :-)
